From bfc5511830458aa8b210d23c018968934887f5c9 Mon Sep 17 00:00:00 2001
From: JP Abgrall <jpa@google.com>
Date: Tue, 17 Jun 2014 17:01:14 -0700
Subject: [PATCH 1/4] F2FS support for fastboot format and fsmgr

- Enables fsck.f2fs for partitions with type F2FS.
- Adds support for f2fs as format fs-type.
- f2fs is optional and support is dlopen()d when needed.

Needs changes in
  system/extras/f2fs_utils
  external/f2fs-tools

Change-Id: I8d141a0d4d14df9fe84d3b131484e9696fcd8870
Signed-off-by: JP Abgrall <jpa@google.com>
(cherry picked from commit 1235158c29909132fbe9aa52939061fc81f0800e)

Backport only the fs_mgr filesystem-check support to CM11. Omit the
fastboot formatter changes, which depend on later f2fs_utils host libraries
and are not needed when formatting partitions in recovery.

Change-Id: I6838c15f4e24a2df710fa654a120a72b9d58ee82
---
 fs_mgr/fs_mgr.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/fs_mgr/fs_mgr.c b/fs_mgr/fs_mgr.c
index 006560b..ece2301 100644
--- a/fs_mgr/fs_mgr.c
+++ b/fs_mgr/fs_mgr.c
@@ -51,6 +51,7 @@
 #define KEY_IN_FOOTER  "footer"
 
 #define E2FSCK_BIN      "/system/bin/e2fsck"
+#define F2FS_FSCK_BIN  "/system/bin/fsck.f2fs"
 #define MKSWAP_BIN      "/system/bin/mkswap"
 
 #define FSCK_LOG_FILE   "/dev/fscklogs/log"
@@ -444,6 +445,20 @@ static void check_fs(char *blk_device, char *fs_type, char *target)
             /* No need to check for error in fork, we can't really handle it now */
             ERROR("Failed trying to run %s\n", E2FSCK_BIN);
         }
+    } else if (!strcmp(fs_type, "f2fs")) {
+            char *f2fs_fsck_argv[] = {
+                    F2FS_FSCK_BIN,
+                    blk_device
+            };
+        INFO("Running %s on %s\n", F2FS_FSCK_BIN, blk_device);
+
+        ret = android_fork_execvp_ext(ARRAY_SIZE(f2fs_fsck_argv), f2fs_fsck_argv,
+                                      &status, true, LOG_KLOG | LOG_FILE,
+                                      true, FSCK_LOG_FILE);
+        if (ret < 0) {
+            /* No need to check for error in fork, we can't really handle it now */
+            ERROR("Failed trying to run %s\n", F2FS_FSCK_BIN);
+        }
     }
 
     return;
